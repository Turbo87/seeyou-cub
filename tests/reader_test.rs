use seeyou_cub::{CubReader, CubStyle};

#[test]
fn parse_france_fixture() {
    let mut reader = CubReader::from_path("tests/fixtures/france_2024.07.02.cub")
        .expect("Failed to open fixture file");
    let mut warnings = Vec::new();

    let header = reader
        .read_header(&mut warnings)
        .expect("Failed to parse header");

    // Validate warnings
    assert_eq!(warnings.len(), 0);

    // Snapshot the header
    insta::assert_debug_snapshot!("header", header);

    // Parse all items
    let items: Vec<_> = reader
        .read_items(&header, &mut warnings)
        .expect("Failed to create item iterator")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse items");

    // Validate item count
    assert_eq!(items.len(), 1368);

    // Snapshot first few items
    insta::assert_debug_snapshot!("items_sample", &items[0..5]);

    // Validate first item's points
    let first_item = items.first().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, first_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("first_item_points", points);

    let last_item = items.last().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, last_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("last_item_points", points);

    // Validate all airspaces and collect statistics
    let mut total_points = 0;
    let mut style_counts = std::collections::HashMap::new();

    for item in &items {
        *style_counts.entry(item.style()).or_insert(0) += 1;

        let mut point_iter = reader
            .read_points(&header, item, &mut warnings)
            .expect("Failed to read points");

        for point_result in &mut point_iter {
            let _point = point_result.expect("Failed to parse point");
            total_points += 1;
        }
    }

    // Validate total points
    assert_eq!(total_points, 93784, "Total points mismatch");

    // Validate airspace type distribution
    assert_eq!(
        *style_counts.get(&CubStyle::Unknown).unwrap_or(&0),
        452,
        "Unknown airspace count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::RestrictedArea).unwrap_or(&0),
        435,
        "Restricted area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ProhibitedArea).unwrap_or(&0),
        113,
        "Prohibited area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::DangerArea).unwrap_or(&0),
        111,
        "Danger area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::GliderSector).unwrap_or(&0),
        135,
        "Glider sector count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ControlZone).unwrap_or(&0),
        92,
        "Control zone count mismatch"
    );
}
