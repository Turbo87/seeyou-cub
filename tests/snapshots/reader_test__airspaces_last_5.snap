---
source: tests/reader_test.rs
expression: "&airspaces[airspaces.len() - 5..]"
---
[
    Airspace {
        left: 0.020236123353242874,
        top: 0.8683302998542786,
        right: 0.06833907216787338,
        bottom: 0.8377871513366699,
        style: Unknown,
        class: ClassA,
        extended_type: None,
        min_alt: 1981,
        max_alt: 2591,
        min_alt_style: FlightLevel,
        max_alt_style: FlightLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.868301510810852, lon: 0.02443445846438408 },
            Point { lat: 0.8651013970375061, lon: 0.022398144006729126 },
            Point { lat: 0.8592838048934937, lon: 0.021525666117668152 },
            Point { lat: 0.8581199645996094, lon: 0.022980328649282455 },
            Point { lat: 0.8576304316520691, lon: 0.025981269776821136 },
            Point { lat: 0.8569758534431458, lon: 0.026097387075424194 },
            Point { lat: 0.8537666201591492, lon: 0.02476896159350872 },
            Point { lat: 0.8500769734382629, lon: 0.024876024574041367 },
            Point { lat: 0.8492289781570435, lon: 0.022446613758802414 },
            Point { lat: 0.8482058048248291, lon: 0.021079838275909424 },
            Point { lat: 0.8473088145256042, lon: 0.020236123353242874 },
            Point { lat: 0.8392173051834106, lon: 0.021040955558419228 },
            Point { lat: 0.8377871513366699, lon: 0.024487722665071487 },
            Point { lat: 0.8379133939743042, lon: 0.02627209573984146 },
            Point { lat: 0.8380199670791626, lon: 0.02783328853547573 },
            Point { lat: 0.8386213183403015, lon: 0.029830720275640488 },
            Point { lat: 0.838805079460144, lon: 0.030140720307826996 },
            Point { lat: 0.8393435478210449, lon: 0.030591875314712524 },
            Point { lat: 0.8429453372955322, lon: 0.03311289846897125 },
            Point { lat: 0.8425335884094238, lon: 0.050730783492326736 },
            Point { lat: 0.8395278453826904, lon: 0.05330028012394905 },
            Point { lat: 0.8393142819404602, lon: 0.05655848607420921 },
            Point { lat: 0.8448458313941956, lon: 0.06598316878080368 },
            Point { lat: 0.847250759601593, lon: 0.06833907216787338 },
            Point { lat: 0.8489184379577637, lon: 0.06724874675273895 },
            Point { lat: 0.8494905233383179, lon: 0.06600234657526016 },
            Point { lat: 0.849694013595581, lon: 0.06510057300329208 },
            Point { lat: 0.8499848246574402, lon: 0.06381582468748093 },
            Point { lat: 0.8506155014038086, lon: 0.06382594257593155 },
            Point { lat: 0.8570584058761597, lon: 0.06390797346830368 },
            Point { lat: 0.8596912622451782, lon: 0.06776221841573715 },
            Point { lat: 0.8649170994758606, lon: 0.06572110950946808 },
            Point { lat: 0.8682722449302673, lon: 0.06388400495052338 },
            Point { lat: 0.8676857948303223, lon: 0.06081967428326607 },
            Point { lat: 0.8668468594551086, lon: 0.05648551508784294 },
            Point { lat: 0.8658918142318726, lon: 0.05420684441924095 },
            Point { lat: 0.8657267093658447, lon: 0.051947884261608124 },
            Point { lat: 0.8669779300689697, lon: 0.048568759113550186 },
            Point { lat: 0.8662993311882019, lon: 0.04788963496685028 },
            Point { lat: 0.8641223907470703, lon: 0.04727922007441521 },
            Point { lat: 0.8634384870529175, lon: 0.04542719945311546 },
            Point { lat: 0.8627742528915405, lon: 0.04363324120640755 },
            Point { lat: 0.8645825982093811, lon: 0.02907436154782772 },
            Point { lat: 0.86622154712677, lon: 0.030087457969784737 },
            Point { lat: 0.8683302998542786, lon: 0.026810072362422943 },
            Point { lat: 0.868301510810852, lon: 0.024434460327029228 },
        ],
        name: Some(
            "Paris7",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        left: -0.002758589806035161,
        top: 0.8217399716377258,
        right: 0.05109962821006775,
        bottom: 0.7809960842132568,
        style: Unknown,
        class: ClassD,
        extended_type: None,
        min_alt: 3505,
        max_alt: 4420,
        min_alt_style: FlightLevel,
        max_alt_style: FlightLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.8217400312423706, lon: 0.037854257971048355 },
            Point { lat: 0.8122082352638245, lon: 0.028943583369255066 },
            Point { lat: 0.8130275011062622, lon: 0.02025555446743965 },
            Point { lat: 0.8101235032081604, lon: 0.020216669887304306 },
            Point { lat: 0.8104872703552246, lon: 0.01600874774158001 },
            Point { lat: 0.7996516227722168, lon: 0.0009454488754272461 },
            Point { lat: 0.7997437715530396, lon: -0.0012751631438732147 },
            Point { lat: 0.7975135445594788, lon: -0.0027585886418819427 },
            Point { lat: 0.7975140810012817, lon: -0.0027564577758312225 },
            Point { lat: 0.7974538803100586, lon: -0.0027303583920001984 },
            Point { lat: 0.7973851561546326, lon: -0.002703726291656494 },
            Point { lat: 0.7973158955574036, lon: -0.002680288627743721 },
            Point { lat: 0.7972456216812134, lon: -0.0026611145585775375 },
            Point { lat: 0.7971752882003784, lon: -0.002644602209329605 },
            Point { lat: 0.7971044778823853, lon: -0.002631818875670433 },
            Point { lat: 0.7970336079597473, lon: -0.002622762694954872 },
            Point { lat: 0.7970336079597473, lon: -0.002622762694954872 },
            Point { lat: 0.7941248416900635, lon: 0.0016437508165836334 },
            Point { lat: 0.7941248416900635, lon: 0.01815159246325493 },
            Point { lat: 0.7844913601875305, lon: 0.022373896092176437 },
            Point { lat: 0.7837786674499512, lon: 0.01868477277457714 },
            Point { lat: 0.7809960842132568, lon: 0.018461594358086586 },
            Point { lat: 0.781563401222229, lon: 0.0257871076464653 },
            Point { lat: 0.78119957447052, lon: 0.03490658104419708 },
            Point { lat: 0.7909054756164551, lon: 0.03490658104419708 },
            Point { lat: 0.7930099964141846, lon: 0.04363349452614784 },
            Point { lat: 0.8068562150001526, lon: 0.04363349452614784 },
            Point { lat: 0.8073845505714417, lon: 0.04518456757068634 },
            Point { lat: 0.8073115944862366, lon: 0.05109962821006775 },
            Point { lat: 0.8086693286895752, lon: 0.05090520903468132 },
            Point { lat: 0.815941572189331, lon: 0.0494510792195797 },
            Point { lat: 0.8217400312423706, lon: 0.037854257971048355 },
        ],
        name: Some(
            "CTA Limoges 119.200",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: Some(
            119200,
        ),
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        left: 0.010631963610649109,
        top: 0.8698092699050903,
        right: 0.07419077306985855,
        bottom: 0.8351109623908997,
        style: Unknown,
        class: ClassA,
        extended_type: None,
        min_alt: 2591,
        max_alt: 3505,
        min_alt_style: FlightLevel,
        max_alt_style: FlightLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.8698093295097351, lon: 0.02450266107916832 },
            Point { lat: 0.8697752356529236, lon: 0.02191345766186714 },
            Point { lat: 0.8592500686645508, lon: 0.015925956889986992 },
            Point { lat: 0.858120322227478, lon: 0.022980352863669395 },
            Point { lat: 0.8576308488845825, lon: 0.025981295853853226 },
            Point { lat: 0.8569762110710144, lon: 0.026097413152456284 },
            Point { lat: 0.8537664413452148, lon: 0.02476898580789566 },
            Point { lat: 0.8500773310661316, lon: 0.024876048788428307 },
            Point { lat: 0.8492288589477539, lon: 0.022446639835834503 },
            Point { lat: 0.8482056260108948, lon: 0.021079864352941513 },
            Point { lat: 0.8473086357116699, lon: 0.020236149430274963 },
            Point { lat: 0.8448898792266846, lon: 0.013191872276365757 },
            Point { lat: 0.8424413204193115, lon: 0.010631963610649109 },
            Point { lat: 0.8365944027900696, lon: 0.012929809279739857 },
            Point { lat: 0.8351109623908997, lon: 0.01792818307876587 },
            Point { lat: 0.8360905051231384, lon: 0.025123197585344315 },
            Point { lat: 0.8379132747650146, lon: 0.0262721199542284 },
            Point { lat: 0.8380197882652283, lon: 0.02783331274986267 },
            Point { lat: 0.8386211395263672, lon: 0.029830744490027428 },
            Point { lat: 0.8388054370880127, lon: 0.030140746384859085 },
            Point { lat: 0.8393434286117554, lon: 0.030591899529099464 },
            Point { lat: 0.8429456949234009, lon: 0.03311292454600334 },
            Point { lat: 0.8425334692001343, lon: 0.05073080584406853 },
            Point { lat: 0.8395277261734009, lon: 0.05330030247569084 },
            Point { lat: 0.8358476161956787, lon: 0.05731487274169922 },
            Point { lat: 0.8394984006881714, lon: 0.06623034179210663 },
            Point { lat: 0.8457043170928955, lon: 0.0712239220738411 },
            Point { lat: 0.8472505807876587, lon: 0.06833910197019577 },
            Point { lat: 0.8489183187484741, lon: 0.06724876910448074 },
            Point { lat: 0.8494903445243835, lon: 0.06600236892700195 },
            Point { lat: 0.8496943712234497, lon: 0.06510059535503387 },
            Point { lat: 0.8499851822853088, lon: 0.06381584703922272 },
            Point { lat: 0.8506153225898743, lon: 0.06382597237825394 },
            Point { lat: 0.8570582270622253, lon: 0.06390799582004547 },
            Point { lat: 0.8596910834312439, lon: 0.06776224076747894 },
            Point { lat: 0.8640300631523132, lon: 0.07419078052043915 },
            Point { lat: 0.8691738247871399, lon: 0.06864963471889496 },
            Point { lat: 0.868272066116333, lon: 0.06388402730226517 },
            Point { lat: 0.8668467402458191, lon: 0.05648554116487503 },
            Point { lat: 0.865891695022583, lon: 0.05420687049627304 },
            Point { lat: 0.8657270669937134, lon: 0.05194791033864021 },
            Point { lat: 0.8669777512550354, lon: 0.048568785190582275 },
            Point { lat: 0.8662991523742676, lon: 0.04788966104388237 },
            Point { lat: 0.864122211933136, lon: 0.047279246151447296 },
            Point { lat: 0.8634388446807861, lon: 0.04542722925543785 },
            Point { lat: 0.8627746105194092, lon: 0.043633267283439636 },
            Point { lat: 0.8645829558372498, lon: 0.02907438762485981 },
            Point { lat: 0.8662213683128357, lon: 0.030087485909461975 },
            Point { lat: 0.8683306574821472, lon: 0.026810098439455032 },
            Point { lat: 0.8698093295097351, lon: 0.024502666667103767 },
        ],
        name: Some(
            "Paris9",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        left: 0.013671745546162128,
        top: 0.8726646304130554,
        right: 0.06952206790447235,
        bottom: 0.8377580642700195,
        style: Unknown,
        class: ClassE,
        extended_type: None,
        min_alt: 1067,
        max_alt: 3505,
        min_alt_style: MeanSeaLevel,
        max_alt_style: FlightLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.8726646900177002, lon: 0.017453549429774284 },
            Point { lat: 0.8464844822883606, lon: 0.017453549429774284 },
            Point { lat: 0.8417673110961914, lon: 0.013671745546162128 },
            Point { lat: 0.8417673110961914, lon: 0.02379952371120453 },
            Point { lat: 0.840376079082489, lon: 0.02844421938061714 },
            Point { lat: 0.840376079082489, lon: 0.03577931970357895 },
            Point { lat: 0.8377580642700195, lon: 0.043633222579956055 },
            Point { lat: 0.8377580642700195, lon: 0.06952206790447235 },
            Point { lat: 0.8443853259086609, lon: 0.06635653227567673 },
            Point { lat: 0.8476626873016357, lon: 0.06477084010839462 },
            Point { lat: 0.8557930588722229, lon: 0.06079568713903427 },
            Point { lat: 0.8607381582260132, lon: 0.058870695531368256 },
            Point { lat: 0.8624932169914246, lon: 0.058206479996442795 },
            Point { lat: 0.8668470978736877, lon: 0.05648549273610115 },
            Point { lat: 0.8705218434333801, lon: 0.055161863565444946 },
            Point { lat: 0.8726646900177002, lon: 0.05439591407775879 },
            Point { lat: 0.8726646900177002, lon: 0.017453547567129135 },
        ],
        name: Some(
            "Paris1",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        left: 0.013671745546162128,
        top: 0.8726646304130554,
        right: 0.06952206790447235,
        bottom: 0.8377580642700195,
        style: Unknown,
        class: ClassD,
        extended_type: None,
        min_alt: 3505,
        max_alt: 5944,
        min_alt_style: FlightLevel,
        max_alt_style: FlightLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.8726646900177002, lon: 0.017453549429774284 },
            Point { lat: 0.8464844822883606, lon: 0.017453549429774284 },
            Point { lat: 0.8417673110961914, lon: 0.013671745546162128 },
            Point { lat: 0.8417673110961914, lon: 0.02379952371120453 },
            Point { lat: 0.840376079082489, lon: 0.02844421938061714 },
            Point { lat: 0.840376079082489, lon: 0.03577931970357895 },
            Point { lat: 0.8377580642700195, lon: 0.043633222579956055 },
            Point { lat: 0.8377580642700195, lon: 0.06952206790447235 },
            Point { lat: 0.8443853259086609, lon: 0.06635653227567673 },
            Point { lat: 0.8476626873016357, lon: 0.06477084010839462 },
            Point { lat: 0.8557930588722229, lon: 0.06079568713903427 },
            Point { lat: 0.8607381582260132, lon: 0.058870695531368256 },
            Point { lat: 0.8624932169914246, lon: 0.058206479996442795 },
            Point { lat: 0.8668470978736877, lon: 0.05648549273610115 },
            Point { lat: 0.8705218434333801, lon: 0.055161863565444946 },
            Point { lat: 0.8726646900177002, lon: 0.05439591407775879 },
            Point { lat: 0.8726646900177002, lon: 0.017453547567129135 },
        ],
        name: Some(
            "Paris1",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
]
